src/lara-app/components/runtime.tsx

```diff
@@ -79,34 +84,50 @@ export const RuntimeComponent = ({experiment, runKey, firebaseJWT, setError, def
 
   // get experiment data when loaded
   useEffect(() => {
-    firebase
-      .firestore()
-      .collection(`runs/${runKey}/experiments`)
-      .orderBy("createdAt", "desc")
-      .limit(1)
-      .onSnapshot(snapshot => {
-        setQueriedFirestore(true);
-        experimentRef.current = snapshot.docs[0]?.data?.().experiment;
-        const newData = experimentRef.current?.data as IExperimentData | undefined;
-        setExperimentData(newData);
-
-        // if there is no data force an upload
-        setDisplayQrAndMaybeRegenerateQR((newData === undefined) && !reportMode);
-      }, (err) => {
-        setError(err);
-      });
+    if (runKey) {
+      firebase
+        .firestore()
+        .collection(`runs/${runKey}/experiments`)
+        .orderBy("createdAt", "desc")
+        .limit(1)
+        .onSnapshot(snapshot => {
+          setQueriedFirestore(true);
+          experimentRef.current = snapshot.docs[0]?.data?.().experiment;
+          const newData = experimentRef.current?.data as IExperimentData | undefined;
+          setExperimentData(newData);
+
+          // if there is no data force an upload
+          setDisplayQrAndMaybeRegenerateQR((newData === undefined) && !reportOrPreviewMode);
+        }, (err) => {
+          setError(err);
+        });
+    }
```

Comment:
Not sure if it's causing any issues in practice, probably not - but it might not be fully correct logic if you consider how React hooks work. `useEffect` is called with `[]`, but the callback strongly depends on `runKey` and some other variables (e.g. `reportOrPreviewMode`). So, if `runKey` gets updated later (e.g. cause it's finally sent from LARA), this function won't run again and won't download some data from Firestone.

Reply:
Consider adding [tslint-react-hooks](https://www.npmjs.com/package/tslint-react-hooks) to the project, which reports missing hook dependencies as lint errors.

Reply:
This `useEffect` is basically a `componentDidMount` and should only fire once when this component is used (which is after we get the callback from Lara).  The runKey doesn't change and is false in this new preview mode so there is no need to re-run the effect.

Reply:
I've added a comment about this in the code.

Reply:
I fully agree it's not causing any issues in practice right now and 99% won't cause any problems in the future.

But... :wink: at the same time, I think it's just not a very good pattern that can cause some hard to detect bugs. That way, we're making assumptions about parent component behavior. What if `runKey` is received with some delay later and it's undefined for a bit? It seems possible given it's received via Interactive API or from some other location. And I wouldn't be able to tell if it's true or not just by looking at `Runtime` class.

Basically, I think that in "reactive" programming we shouldn't make any assumptions about parent behavior. And I guess we shouldn't assume that some properties are not going to change (since they're props). Just "react" to current values of props and state.

I know it's pretty theoretical and nitpicking, but I'm mostly mentioning that, as I did **exactly** the same thing myself here:
https://github.com/concord-consortium/question-interactives/tree/598de57492d7fa9e69f6194f0b44f5424422790b
where `disabled` is not listed as a dependency. It was not causing any problems until Chris made some changes in the parent component and I believe it gave her some headaches. :wink: disabled was `undefined` for a bit, then it was turning to `false`, but autoHeight was never disabled. This is at least my understanding of how these hooks work. And I think this assumption that `useEffect(, []) === componentDidMount` can be risky because of that.